pytest_redis.py

```diff
@@ -54,19 +80,105 @@ def pytest_collection(session, genitems=True):
+    If it is not empty, then pop all elements off and push them onto
+    the main redis list and then return a generator for that list.
+
+    Finally, if both lists are empty, return the main redis list
+    iterator as is.
+    """
+    redis_list_key = session.config.getoption("redis_list_key")
+    ran_test_list_key = session.config.getoption("redis_backup_list_key")
+
+    final_redis_list = None
+
+    main_redis_list = redis_test_generator(session.config, redis_connection,
+                                           redis_list_key, redis_pop_type)
+
+    main_list_next_val, main_redis_list = peek_next_item(main_redis_list)
+
+    if main_list_next_val is None and ran_test_list_key is not None:
+        prev_list = redis_test_generator(session.config,
+                                         redis_connection,
+                                         ran_test_list_key,
+                                         redis_pop_type)
+        # Peek at the backup redis list
+        prev_list_next_val, prev_list = peek_next_item(prev_list)
+
+        # Check if the backup redist list is empty
+        if prev_list_next_val is not None:
+            # Push tests to the main redis list
+            for test in prev_list:
+                push_tests_to_redis(redis_connection, redis_list_key,
+                                    redis_push_type, test)
+        # Regenerate the final redis list
+        final_redis_list = redis_test_generator(session.config,
+                                                redis_connection,
+                                                redis_list_key,
+                                                redis_pop_type)
+    else:
+        final_redis_list = main_redis_list
+
+    return final_redis_list
+
+
 def perform_collect_and_run(session):
     """Collect and run tests streaming from the redis queue."""
     # This mimics the internal pytest collect loop, but shortened
     # while running tests as soon as they are found.
     term = TerminalReporter(session.config)
-    redis_list = redis_test_generator(session.config, session.config.args)
+
+    redis_connection = get_redis_connection(session.config)
+    redis_push_type = session.config.getoption('redis_push_type')
+    redis_pop_type = session.config.getoption('redis_pop_type')
+
+    redis_list = determine_redis_list_key_to_use(session,
+                                                 redis_connection,
+                                                 redis_pop_type,
+                                                 redis_push_type)
+
+    ran_test_list_key = session.config.getoption("redis_backup_list_key")
+
     hook = session.config.hook
     session._initialpaths = set()
     session._initialparts = []
     session._notfound = []
     session.items = []
     for arg in redis_list:
         term.write(os.linesep)
+        if ran_test_list_key:
+            push_tests_to_redis(redis_connection, ran_test_list_key,
```

Comment:
I think it might be clearer to say that if there is a `backup_list_key` provided then push the test to the redis list -- Otherwise, we would need to put the `RPOPLPUSH` inside `retrieve_test_from_redis`